File: CPP/No567.cc
/**
 * Created by Xiaozhong on 2020/10/20.
 * Copyright (c) 2020/10/20 Xiaozhong. All rights reserved.
 */
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
private:
    inline bool match(const vector<int> &s1map, const vector<int> &s2map) {
        for (int i = 0; i < 26; ++i) {
            if (s1map[i] != s2map[i]) return false;
        }
        return true;
    }

public:
    bool checkInclusion(string s1, string s2) {
        if (s1.length() > s2.length()) return false;
        vector<int> s1map(26);
        for (const char &c : s1) s1map[c - 'a']++;

        for (int i = 0; i <= s2.length() - s1.length(); ++i) {
            vector<int> s2map(26);
            for (int j = 0; j < s1.length(); ++j) {
                s2map[s2[i + j] - 'a']++;
            }
            if (match(s1map, s2map)) return true;
        }
        return false;
    }
};

int main() {
    string s1 = "ab";
    string s2 = "eidbaooo";
    Solution s;
    cout << s.checkInclusion(s1, s2) << endl;
    s1 = "adc";
    s2 = "dcda";
    cout << s.checkInclusion(s1, s2);
}

File: Java/src/No567.java
import java.util.Arrays;

/**
 * Created by Xiaozhong on 2021/2/10.
 * Copyright (c) 2021/2/10 Xiaozhong. All rights reserved.
 */

public class No567 {
    public boolean checkInclusion(String s1, String s2) {
        int n1 = s1.length(), n2 = s2.length();
        if (n1 > n2) return false;
        int[] cnt1 = new int[26], cnt2 = new int[26];
        for (int i = 0; i < n1; ++i) {
            cnt1[s1.charAt(i) - 'a']++;
            cnt2[s2.charAt(i) - 'a']++;
        }
        if (Arrays.equals(cnt1, cnt2)) return true;
        for (int i = n1; i < n2; ++i) {
            cnt2[s2.charAt(i - n1) - 'a']--;
            cnt2[s2.charAt(i) - 'a']++;
            if (Arrays.equals(cnt1, cnt2)) return true;
        }
        return false;
    }
}
